### PR TITLE
CA-228029: Fix update destroy failure in xe-install-supplemental-pack when exception happens

### PR DESCRIPTION
**Pull request by cheng-z**
The problem here is that for code like
trap "xe update-pool-clean uuid=$update_uuid" EXIT ERR
trap "xe update-destroy uuid=$update_uuid" ERR

In testing, it will call xe update-destroy first.
After xe update-destroy, xe update-pool-clean will have no object to refer.

We expect the code will call pool_clean first then destroy.
But what we wrote in trap seems cannot make sure this sequence

Signed-off-by: Cheng Zhang <cheng.zhang@citrix.com>

### CODE REVIEW

**Comment by cheng-z** [comment]
Hi @srowe 
Could you please also help to review this code?
Thanks.

**Comment by lindig** [comment]
This needs a better commit message about what problem is fixed here. They way I understand it the clean-up code is now called directly rather than via a trap. Why did the previous solution not work and why is this working now?

File: scripts/xe-install-supplemental-pack [hunk 1]
```diff
@@ -32,6 +32,6 @@ iso="$1"
 [ -n "$iso" ] || usage
 
 update_uuid=$(xe update-upload file-name="$iso" ${sr_uuid:+sr-uuid=$sr_uuid})
-trap "xe update-pool-clean uuid=$update_uuid" EXIT ERR
-trap "xe update-destroy uuid=$update_uuid" ERR
+trap "xe update-pool-clean uuid=$update_uuid; xe update-destroy uuid=$update_uuid" ERR
```
Comment (srowe) [inline scripts/xe-install-supplemental-pack]: If the `update-pool-clean` fails then the script will terminate without calling `update-destroy`. You can fix this by having `set +e` first.

**Comment by cheng-z** [comment]
Hi @srowe 
I modify the code by add "set +e" at the begin and "set -e" at the end.
Please help to review the again. Thanks a lot.

**Comment by cheng-z** [comment]
Hi @lindig 
Comments updated. Thanks a lot for your comments.

**Comment by srowe** [comment]
I'm not clear why the original trap handling was wrong.

**Comment by srowe** [comment]
I'll follow up on the ticket

**Comment by cheng-z** [comment]
Hi @srowe 
The problem here is that for code like
trap "xe update-pool-clean uuid=$update_uuid" EXIT ERR
trap "xe update-destroy uuid=$update_uuid" ERR

In testing, it will call xe update-destroy first.
After xe update-destroy, xe update-pool-clean will have no object to refer.

We expect the code will call pool_clean first then destroy.

**Comment by lindig** [comment]
I think the last comment above should go into the commit message. The discussion here shows that the change is not obvious.

**Comment by huizh** [comment]
Hi @cheng-z , I think this commit only needs to change `scripts/xe-install-supplemental-pack`, another one `ocaml/xapi/cli_operations.ml` has been merged with https://github.com/xapi-project/xen-api/pull/2821/files, right?

Thanks.

**Comment by jonludlam** [comment]
You need to remove the livepatch cli stuff - it's conflicting.

**Comment by cheng-z** [comment]
@huizh and @jonludlam 
Updated. Sorry for the rush.

**Comment by lindig** [comment]
Looks good to me. @huizh, @jonludlam, @srowe?

**Comment by cheng-z** [comment]
Hi  @huizh, @jonludlam, @srowe and @lindig 
I added another commit to handle install update multiple time issue.
Will not call update-destroy if the update already installed.

File: scripts/xe-install-supplemental-pack [hunk 2]
```diff
@@ -31,7 +31,17 @@ done
 iso="$1"
 [ -n "$iso" ] || usage
 
+function clean_up {
+  set +e
+  xe update-pool-clean uuid=$update_uuid
+  hosts=$(xe update-param-get param-name=hosts uuid=$update_uuid)
+  if [ $hosts = "" ]; then
```
Comment (cheng-z) [inline scripts/xe-install-supplemental-pack]: Hi @lindig 
Thanks a lot. Updated.

Comment (lindig) [inline scripts/xe-install-supplemental-pack]: This was suggested by @srowe. I'm happy to merge it.

Comment (cheng-z) [inline scripts/xe-install-supplemental-pack]: Oh. my bad. @srowe apology and thanks. :-)

**Pull request merged**